From 500af94404759006b240eaf019bf67f4e22b3da2 Mon Sep 17 00:00:00 2001
From: "guangji.huang" <guangji.huang@sophgo.com>
Date: Wed, 13 May 2026 15:27:12 +0800
Subject: [PATCH] [feat](cvi_mpi): sync teaisp bnr code [description ]:NA [root
   cause]:NA [JIRA      ID]:NA [chip project]:cv184x [side effects]:NA
 [Affected *.ko/*.so/*.a]:NA

Change-Id: Ieca9b60f57d173556fa00356c5aa1de6d2029a17
---
 modules/vi/api/api_vi.c                   |  12 ++
 modules/vi/include/platform_vi.h          |   3 +
 modules/vi/platform/dual_os/platform_vi.c |  42 +++++
 modules/vi/platform/linux/platform_vi.c   |  54 +++++++
 modules/vi/platform/linux/vi_ioctl.c      |  78 +++++++++
 modules/vi/platform/linux/vi_ioctl.h      |   5 +-
 pkgconfig/cvi_common.pc                   |   2 +-
 release.sh                                |   9 +-
 sample_app/sample_ir_auto/Makefile        |   4 +-
 sample_app/sbm/Makefile                   |   6 +-
 sample_app/sensor/Makefile                |   2 +-
 sample_app/vio/Makefile                   |   4 +-
 self_test/vi_ut/Makefile                  |   5 +-
 self_test/vi_ut/vi_ut_comm.c              | 185 ++++++++++++++++++++++
 self_test/vi_ut/vi_ut_comm.h              |   1 +
 self_test/vi_ut/vi_ut_main.c              |  21 +++
 16 files changed, 412 insertions(+), 21 deletions(-)

diff --git a/modules/vi/api/api_vi.c b/modules/vi/api/api_vi.c
index 31678eb..6a9455a 100644
--- a/modules/vi/api/api_vi.c
+++ b/modules/vi/api/api_vi.c
@@ -743,4 +743,16 @@ CVI_S32 CVI_VI_GetDevRxFrameCount(VI_DEV ViDev, CVI_U32 *pu32FrameCount)
 	CHECK_VI_NULL_PTR(pu32FrameCount);
 
 	return platform_vi_getdevrxframecount(ViDev, pu32FrameCount);
+}
+
+CVI_S32 CVI_VI_AiIspCfg(VI_AI_ISP_CFG_S *pstAiIspCfg)
+{
+	CHECK_VI_NULL_PTR(pstAiIspCfg);
+	return platform_vi_aiispcfg(pstAiIspCfg);
+}
+
+CVI_S32 CVI_VI_AiIspInfo(VI_AI_ISP_INFO_WRAP_S *pstAiIspInfoWrap)
+{
+	CHECK_VI_NULL_PTR(pstAiIspInfoWrap);
+	return platform_vi_aiispinfo(pstAiIspInfoWrap);
 }
\ No newline at end of file
diff --git a/modules/vi/include/platform_vi.h b/modules/vi/include/platform_vi.h
index ffe933a..68103b0 100644
--- a/modules/vi/include/platform_vi.h
+++ b/modules/vi/include/platform_vi.h
@@ -80,6 +80,9 @@ CVI_S32 platform_vi_getextchnattr(VI_PIPE ViPipe, VI_CHN ViChn, VI_EXT_CHN_ATTR_
 CVI_S32 platform_vi_setbypassfrm(CVI_U32 snr_num, CVI_U8 bypass_num);
 CVI_S32 platform_vi_setdevrxframecount(VI_DEV ViDev, CVI_U32 u32RxFrameCount);
 CVI_S32 platform_vi_getdevrxframecount(VI_DEV ViDev, CVI_U32 *pu32FrameCount);
+CVI_S32 platform_vi_aiispcfg(VI_AI_ISP_CFG_S *pstAiIspCfg);
+CVI_S32 platform_vi_aiispinfo(VI_AI_ISP_INFO_WRAP_S *pstAiIspInfoWrap);
+
 #ifdef __cplusplus
 #if __cplusplus
 }
diff --git a/modules/vi/platform/dual_os/platform_vi.c b/modules/vi/platform/dual_os/platform_vi.c
index 73ef6c5..c284831 100644
--- a/modules/vi/platform/dual_os/platform_vi.c
+++ b/modules/vi/platform/dual_os/platform_vi.c
@@ -1193,3 +1193,45 @@ CVI_S32 platform_vi_getmipibinddev(VI_DEV ViDev, MIPI_DEV *pMipiDev)
 
 	return CVI_ERR_VI_NOT_SUPPORT;
 }
+
+CVI_S32 platform_vi_aiispcfg(VI_AI_ISP_CFG_S *pstAiIspCfg)
+{
+	CVI_S32 s32Ret = CVI_SUCCESS;
+	CVI_U32 u32ModFd = MODFD(CVI_ID_VI, 0, 0);
+
+	if (pstAiIspCfg == CVI_NULL) {
+		CVI_TRACE_VI(CVI_DBG_ERR, "pstAiIspCfg is NULL\n");
+		return CVI_ERR_VI_INVALID_NULL_PTR;
+	}
+
+	s32Ret = CVI_MSG_SendSync(u32ModFd, MSG_CMD_VI_AI_ISP_CFG,
+				  pstAiIspCfg, sizeof(*pstAiIspCfg), NULL);
+	if (s32Ret != CVI_SUCCESS) {
+		CVI_TRACE_VI(CVI_DBG_ERR, "AI_ISP_CFG fail, ViPipe:%d, ret:0x%x\n",
+			pstAiIspCfg->viPipe, s32Ret);
+		return s32Ret;
+	}
+
+	return CVI_SUCCESS;
+}
+
+CVI_S32 platform_vi_aiispinfo(VI_AI_ISP_INFO_WRAP_S *pstAiIspInfoWrap)
+{
+	CVI_S32 s32Ret = CVI_SUCCESS;
+	CVI_U32 u32ModFd = MODFD(CVI_ID_VI, 0, 0);
+
+	if (pstAiIspInfoWrap == CVI_NULL) {
+		CVI_TRACE_VI(CVI_DBG_ERR, "pstAiIspInfoWrap is NULL\n");
+		return CVI_ERR_VI_INVALID_NULL_PTR;
+	}
+
+	s32Ret = CVI_MSG_SendSync(u32ModFd, MSG_CMD_VI_AI_ISP_INFO,
+				  pstAiIspInfoWrap, sizeof(*pstAiIspInfoWrap), NULL);
+	if (s32Ret != CVI_SUCCESS) {
+		CVI_TRACE_VI(CVI_DBG_ERR, "AI_ISP_INFO fail, ViPipe:%d, ret:0x%x\n",
+			pstAiIspInfoWrap->viPipe, s32Ret);
+		return s32Ret;
+	}
+
+	return CVI_SUCCESS;
+}
diff --git a/modules/vi/platform/linux/platform_vi.c b/modules/vi/platform/linux/platform_vi.c
index f2476c9..15caf2c 100644
--- a/modules/vi/platform/linux/platform_vi.c
+++ b/modules/vi/platform/linux/platform_vi.c
@@ -1451,6 +1451,60 @@ CVI_S32 platform_vi_getdevrxframecount(VI_DEV ViDev, CVI_U32 *pu32FrameCount)
 	return s32Ret;
 }
 
+CVI_S32 platform_vi_aiispcfg(VI_AI_ISP_CFG_S *pstAiIspCfg)
+{
+	CVI_S32 s32Ret = CVI_SUCCESS;
+	CVI_S32 fd = -1;
+
+	fd = get_vi_fd();
+
+	if (pstAiIspCfg == CVI_NULL) {
+		CVI_TRACE_VI(CVI_DBG_ERR, "pstAiIspCfg is NULL\n");
+		return CVI_ERR_VI_INVALID_NULL_PTR;
+	}
+
+	s32Ret = vi_sdk_set_ai_isp_cfg(fd, pstAiIspCfg);
+	if (s32Ret != CVI_SUCCESS) {
+		CVI_TRACE_VI(CVI_DBG_ERR, "vi_sdk_set_ai_isp_cfg failed, ViPipe:%d, ret:0x%x\n",
+			pstAiIspCfg->viPipe, s32Ret);
+		return s32Ret;
+	}
+
+	return CVI_SUCCESS;
+}
+
+CVI_S32 platform_vi_aiispinfo(VI_AI_ISP_INFO_WRAP_S *pstAiIspWrapInfo)
+{
+	CVI_S32 s32Ret = CVI_SUCCESS;
+	CVI_S32 fd = -1;
+
+	fd = get_vi_fd();
+
+	if (pstAiIspWrapInfo == CVI_NULL) {
+		CVI_TRACE_VI(CVI_DBG_ERR, "pstAiIspWrapInfo is NULL\n");
+		return CVI_ERR_VI_INVALID_NULL_PTR;
+	}
+
+	if (pstAiIspWrapInfo->enCmd == VI_CMD_GET_AI_ISP_RAW) {
+		s32Ret = vi_sdk_get_ai_isp_raw(fd, pstAiIspWrapInfo);
+		if (s32Ret != CVI_SUCCESS) {
+			CVI_TRACE_VI(CVI_DBG_ERR, "vi_sdk_get_ai_isp_raw failed, ViPipe:%d, ret:0x%x\n",
+				pstAiIspWrapInfo->viPipe, s32Ret);
+			return s32Ret;
+		}
+
+	} else {
+		s32Ret = vi_sdk_put_ai_isp_raw(fd, pstAiIspWrapInfo);
+		if (s32Ret != CVI_SUCCESS) {
+			CVI_TRACE_VI(CVI_DBG_ERR, "vi_sdk_put_ai_isp_raw failed, ViPipe:%d, ret:0x%x\n",
+				pstAiIspWrapInfo->viPipe, s32Ret);
+			return s32Ret;
+		}
+	}
+
+	return CVI_SUCCESS;
+}
+
 /**
  * @deprecated
  */
diff --git a/modules/vi/platform/linux/vi_ioctl.c b/modules/vi/platform/linux/vi_ioctl.c
index 90325db..f0da664 100644
--- a/modules/vi/platform/linux/vi_ioctl.c
+++ b/modules/vi/platform/linux/vi_ioctl.c
@@ -49,6 +49,23 @@ static inline int G_CTRL_PTR(int fd, void *cfg, int size, enum VI_IOCTL _ioctl)
 	return rc;
 }
 
+static inline int S_CTRL_PTR(int fd, void *cfg, int size, enum VI_IOCTL _ioctl)
+{
+	struct vi_ext_control ec1;
+	int rc = 0;
+
+	memset(&ec1, 0, sizeof(ec1));
+	ec1.id = _ioctl;
+	ec1.ptr = cfg;
+	ec1.size = size;
+
+	rc = ioctl(fd, VI_IOC_S_CTRL, &ec1);
+	if (rc < 0) {
+		fprintf(stderr, "VI_IOC_S_CTRL - %s NG, %s\n", __func__, strerror(errno));
+	}
+	return rc;
+}
+
 static inline int SDK_CTRL_SET_VALUE(int fd, enum VI_SDK_CTRL _ioctl
 					, int dev, int pipe, int chn, int val)
 {
@@ -413,3 +430,64 @@ int vi_sdk_get_dev_rx_frame_count(int fd, int dev, uint32_t *count)
 {
 	return SDK_CTRL_SET_CFG(fd, count, sizeof(uint32_t), VI_SDK_GET_DEV_RX_FRAME_COUNT, dev, -1, -1, -1);
 }
+
+int vi_sdk_set_ai_isp_cfg(int fd, VI_AI_ISP_CFG_S *pstAiIspCfg)
+{
+	struct vi_ai_isp_cfg cfg;
+
+	memset(&cfg, 0, sizeof(cfg));
+	cfg.ViPipe = pstAiIspCfg->viPipe;
+	cfg.ViAiISPType = (CVI_U8)pstAiIspCfg->enAiIspType;
+	cfg.Reserved[0] = pstAiIspCfg->reserved[0];
+	cfg.Reserved[1] = pstAiIspCfg->reserved[1];
+
+	return S_CTRL_PTR(fd, &cfg, sizeof(struct vi_ai_isp_cfg), VI_IOCTL_AI_ISP_CFG);
+}
+
+int vi_sdk_get_ai_isp_raw(int fd, VI_AI_ISP_INFO_WRAP_S *infoWrap)
+{
+	int ret = 0;
+	struct vi_ai_isp_info info;
+
+	memset(&info, 0, sizeof(info));
+
+	info.ViPipe = infoWrap->stIspInfo.viPipe;
+	info.InputAddr[0] = infoWrap->stIspInfo.inputAddr[0];
+	info.InputAddr[1] = infoWrap->stIspInfo.inputAddr[1];
+	info.OutputAddr[0] = infoWrap->stIspInfo.outputAddr[0];
+	info.OutputAddr[1] = infoWrap->stIspInfo.outputAddr[1];
+	info.Size = infoWrap->stIspInfo.size;
+	info.Reserved[0] = infoWrap->stIspInfo.reserved[0];
+
+	ret = G_CTRL_PTR(fd, &info, sizeof(struct vi_ai_isp_info), VI_IOCTL_GET_AI_ISP_RAW);
+	if (ret < 0) {
+		fprintf(stderr, "VI_IOCTL_GET_AI_ISP_RAW - %s NG, %s\n", __func__, strerror(errno));
+		return ret;
+	}
+
+	infoWrap->stIspInfo.inputAddr[0] = info.InputAddr[0];
+	infoWrap->stIspInfo.inputAddr[1] = info.InputAddr[1];
+	infoWrap->stIspInfo.outputAddr[0] = info.OutputAddr[0];
+	infoWrap->stIspInfo.outputAddr[1] = info.OutputAddr[1];
+	infoWrap->stIspInfo.size = info.Size;
+	infoWrap->stIspInfo.reserved[0] = info.Reserved[0];
+
+	return ret;
+}
+
+int vi_sdk_put_ai_isp_raw(int fd, VI_AI_ISP_INFO_WRAP_S *infoWrap)
+{
+	struct vi_ai_isp_info info;
+
+	memset(&info, 0, sizeof(info));
+
+	info.ViPipe = infoWrap->stIspInfo.viPipe;
+	info.InputAddr[0] = infoWrap->stIspInfo.inputAddr[0];
+	info.InputAddr[1] = infoWrap->stIspInfo.inputAddr[1];
+	info.OutputAddr[0] = infoWrap->stIspInfo.outputAddr[0];
+	info.OutputAddr[1] = infoWrap->stIspInfo.outputAddr[1];
+	info.Size = infoWrap->stIspInfo.size;
+	info.Reserved[0] = infoWrap->stIspInfo.reserved[0];
+
+	return S_CTRL_PTR(fd, &info, sizeof(struct vi_ai_isp_info), VI_IOCTL_PUT_AI_ISP_RAW);
+}
\ No newline at end of file
diff --git a/modules/vi/platform/linux/vi_ioctl.h b/modules/vi/platform/linux/vi_ioctl.h
index 87a549c..9a7ce93 100644
--- a/modules/vi/platform/linux/vi_ioctl.h
+++ b/modules/vi/platform/linux/vi_ioctl.h
@@ -58,7 +58,6 @@ int vi_sdk_put_smooth_rawdump(int fd, int pipe, VIDEO_FRAME_INFO_S *pstFrameInfo
 int vi_sdk_set_chn_rotation(int fd, struct vi_chn_rot_cfg *cfg);
 int vi_sdk_get_chn_rotation(int fd, struct vi_chn_rot_cfg *cfg);
 int vi_sdk_set_chn_ldc(int fd, struct vi_chn_ldc_cfg *cfg);
-int vi_sdk_set_chn_ldc(int fd, struct vi_chn_ldc_cfg *cfg);
 int vi_sdk_get_chn_ldc(int fd, int pipe, int chn, struct vi_chn_ldc_cfg *cfg);
 int vi_sdk_set_chn_flip_mirror(int fd, struct vi_chn_flip_mirror_cfg *cfg);
 int vi_sdk_get_chn_flip_mirror(int fd, struct vi_chn_flip_mirror_cfg *cfg);
@@ -68,4 +67,8 @@ int vi_sdk_dump_register(int fd, int pipe, struct ip_info *ip_info);
 int vi_sdk_set_dev_rx_frame_count(int fd, int dev, uint32_t count);
 int vi_sdk_get_dev_rx_frame_count(int fd, int dev, uint32_t *count);
 
+int vi_sdk_set_ai_isp_cfg(int fd, VI_AI_ISP_CFG_S *pstAiIspCfg);
+int vi_sdk_get_ai_isp_raw(int fd, VI_AI_ISP_INFO_WRAP_S *pstAiIspInfoWrap);
+int vi_sdk_put_ai_isp_raw(int fd, VI_AI_ISP_INFO_WRAP_S *pstAiIspInfoWrap);
+
 #endif // MODULES_VPU_INCLUDE_VI_IOCTL_H_
diff --git a/pkgconfig/cvi_common.pc b/pkgconfig/cvi_common.pc
index 38fb650..58c4397 100644
--- a/pkgconfig/cvi_common.pc
+++ b/pkgconfig/cvi_common.pc
@@ -6,4 +6,4 @@ Name: CVITEK System lib
 Description: SYS libs
 Version: 2.0
 Cflags: -I${includedir}
-Libs: -L${libdir} -lsys -lvi -lvpss -lvenc -lvdec -lrgn -lgdc -lisp -lae -laf -lawb -lmipi -lvo
+Libs: -L${libdir} -lsys -lvi -lvpss -lvenc -lvdec -lrgn -lgdc -lisp -lae -laf -lawb -lisp_algo -lmipi -lvo
diff --git a/sample_app/sample_ir_auto/Makefile b/sample_app/sample_ir_auto/Makefile
index b1edca1..a24bfa7 100644
--- a/sample_app/sample_ir_auto/Makefile
+++ b/sample_app/sample_ir_auto/Makefile
@@ -30,9 +30,9 @@ REQUIRES = cvi_common
 LIBS = $(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) pkg-config --libs --define-variable=mw_dir=$(MW_PATH) cvi_common)
 else
 ELFFLAGS += -static
-LIBS = -lvi -lvpss -lvenc -lrgn -lvo -lmipi -lsys -lsensor -lcvi_bin -lisp -lawb -lae -laf
+LIBS = -lvi -lvpss -lvenc -lrgn -lvo -lmipi -lsys -lsensor -lcvi_bin -lisp -lisp_algo -lawb -lae -laf
 ifneq ($(OS_TYPE), DUAL_OS)
-LIBS += -lsns_full -lsensor $(ISP_LIB) -lisp -lawb -lae -laf -lisp_algo -lcvi_bin
+LIBS += -lsns_full -lsensor $(ISP_LIB) -lisp -lawb -lae -laf -lcvi_bin
 endif
 endif
 
diff --git a/sample_app/sbm/Makefile b/sample_app/sbm/Makefile
index a03e428..55b88c8 100644
--- a/sample_app/sbm/Makefile
+++ b/sample_app/sbm/Makefile
@@ -18,11 +18,11 @@ REQUIRES = cvi_common
 LIBS += $(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) pkg-config --libs --define-variable=mw_dir=$(MW_PATH) cvi_common)
 else
 ELFFLAGS += -static
-LIBS +=  -lsensor -lvi -lvpss -lsys -lvenc -lrgn
+LIBS +=  -lsensor -lvi -lvpss -lsys -lvenc -lrgn -lvo -lcvi_bin -lisp -lisp_algo -lawb -lae -laf
 endif
 
 ifneq ($(OS_TYPE), DUAL_OS)
-LIBS += -lsns_full $(ISP_LIB)
+LIBS += -lsns_full
 endif
 
 EXTRA_CFLAGS = $(INCS) $(DEFS)
@@ -40,7 +40,7 @@ $(SDIR)/%.o: $(SDIR)/%.c
 	@echo [$(notdir $(CC))] $(notdir $@)
 
 $(TARGET): $(COMM_OBJ) $(OBJS)
-	@$(CXX) -o $@ -Wl,--start-group $(OBJS) $(COMM_OBJS) -Wl,--end-group $(ELFFLAGS) $(EXTRA_LDFLAGS)
+	@$(CXX) -o $@ -Wl,--start-group $(OBJS) $(COMM_OBJS) $(EXTRA_LDFLAGS) -Wl,--end-group $(ELFFLAGS)
 	@echo -e $(BLUE)[LINK]$(END)[$(notdir $(CXX))] $(notdir $@)
 
 clean:
diff --git a/sample_app/sensor/Makefile b/sample_app/sensor/Makefile
index 83044b3..a7a40fc 100644
--- a/sample_app/sensor/Makefile
+++ b/sample_app/sensor/Makefile
@@ -18,7 +18,7 @@ INCS = -I$(MW_INC) -I$(ISP_INC) -I$(MW_INC)/sensor -I./inc -I$(COMM_INC)
 TARGET = sample_sensor
 
 ifeq ($(OS_TYPE), DUAL_OS)
-LIBS += -lvi -lvpss -lsys -lini -lsensor -lisp -lawb -lae -laf
+LIBS += -lvi -lvpss -lsys -lini -lsensor -lisp -lisp_algo -lawb -lae -laf
 else
 LIBS = -lvi -lini -lsns_full -lsensor -lgdc -lsys -lisp -lcvi_bin -lawb -lae -laf -lisp_algo -lvo -lvpss
 endif
diff --git a/sample_app/vio/Makefile b/sample_app/vio/Makefile
index 1ebd45a..c94fe93 100644
--- a/sample_app/vio/Makefile
+++ b/sample_app/vio/Makefile
@@ -30,9 +30,9 @@ REQUIRES = cvi_common
 LIBS = $(shell PKG_CONFIG_PATH=$(PKG_CONFIG_PATH) pkg-config --libs --define-variable=mw_dir=$(MW_PATH) cvi_common)
 else
 ELFFLAGS += -static
-LIBS = -lvi -lvpss -lvenc -lrgn -lvo -lmipi -lsys -lsensor -lcvi_bin -lisp -lawb -lae -laf
+LIBS = -lvi -lvpss -lvenc -lrgn -lvo -lmipi -lsys -lsensor -lcvi_bin -lisp -lisp_algo -lawb -lae -laf
 ifneq ($(OS_TYPE), DUAL_OS)
-LIBS += -lsns_full -lsensor $(ISP_LIB) -lisp -lawb -lae -laf -lisp_algo -lcvi_bin
+LIBS += -lsns_full -lsensor $(ISP_LIB) -lisp -lawb -lae -laf -lcvi_bin
 endif
 endif
 
diff --git a/self_test/vi_ut/Makefile b/self_test/vi_ut/Makefile
index afac338..aa89272 100644
--- a/self_test/vi_ut/Makefile
+++ b/self_test/vi_ut/Makefile
@@ -16,10 +16,7 @@ else
 	CFLAGS += -DCV184X_FPGA_RAW_REPLAY
 endif
 
-ISP_LIBS = -lae -lawb -laf -lcvi_ispd2 -lisp -ldl
-ifeq ($(OS_TYPE), LINUX)
-	ISP_LIBS += -lisp_algo
-endif
+ISP_LIBS = -lae -lawb -laf -lcvi_ispd2 -lisp -lisp_algo -ldl
 
 OBJS = $(SRCS:.c=.o)
 DEPS = $(SRCS:.c=.d)
diff --git a/self_test/vi_ut/vi_ut_comm.c b/self_test/vi_ut/vi_ut_comm.c
index f747708..b9569b7 100644
--- a/self_test/vi_ut/vi_ut_comm.c
+++ b/self_test/vi_ut/vi_ut_comm.c
@@ -1697,3 +1697,188 @@ CVI_S32 vi_test(VI_UT_CTX *pUtCtx)
 
 	return s32Ret;
 }
+
+#define AI_ISP_TEST_LOOP_CNT (10)
+
+static pthread_t ai_isp_thread_id;
+static volatile CVI_BOOL ai_isp_thread_running = CVI_FALSE;
+
+static void *ai_isp_raw_loop_thread(void *arg)
+{
+	VI_PIPE ViPipe = (VI_PIPE)(uintptr_t)arg;
+	VI_AI_ISP_INFO_WRAP_S wrap_info;
+	CVI_S32 s32Ret;
+	CVI_U32 loop_cnt = 0;
+
+	UT_PRT("[AI_ISP] raw loop thread start, pipe=%d\n", ViPipe);
+
+	prctl(PR_SET_NAME, "ai_isp_raw_loop");
+
+	while (ai_isp_thread_running) {
+		memset(&wrap_info, 0, sizeof(wrap_info));
+		wrap_info.viPipe = ViPipe;
+		wrap_info.enCmd = VI_CMD_GET_AI_ISP_RAW;
+
+		s32Ret = CVI_VI_AiIspInfo(&wrap_info);
+		if (s32Ret != CVI_SUCCESS) {
+			UT_PRT("[AI_ISP] CVI_VI_AiIspInfo GET failed, ret=%#x\n", s32Ret);
+			usleep(5000);
+			continue;
+		}
+
+		if (wrap_info.stIspInfo.inputAddr[0] == 0 || wrap_info.stIspInfo.outputAddr[0] == 0) {
+			UT_PRT("[AI_ISP] get raw addr zero, retry\n");
+			usleep(5000);
+			continue;
+		}
+
+		UT_PRT("[AI_ISP] loop %u: input=0x%llx output=0x%llx size=%u\n",
+			loop_cnt,
+			(unsigned long long)wrap_info.stIspInfo.inputAddr[0],
+			(unsigned long long)wrap_info.stIspInfo.outputAddr[0],
+			wrap_info.stIspInfo.size);
+
+		// bypass TPU: memcpy input raw to output raw
+		{
+			CVI_U32 size = wrap_info.stIspInfo.size;
+			void *src = CVI_SYS_Mmap(wrap_info.stIspInfo.inputAddr[0], size);
+			void *dst = CVI_SYS_Mmap(wrap_info.stIspInfo.outputAddr[0], size);
+
+			if (src && dst) {
+				memcpy(dst, src, size);
+				CVI_SYS_Munmap(src, size);
+				CVI_SYS_Munmap(dst, size);
+			} else {
+				UT_PRT("[AI_ISP] mmap failed\n");
+				if (src)
+					CVI_SYS_Munmap(src, size);
+				if (dst)
+					CVI_SYS_Munmap(dst, size);
+			}
+		}
+
+		memset(&wrap_info, 0, sizeof(wrap_info));
+		wrap_info.viPipe = ViPipe;
+		wrap_info.enCmd = VI_CMD_PUT_AI_ISP_RAW;
+
+		s32Ret = CVI_VI_AiIspInfo(&wrap_info);
+		if (s32Ret != CVI_SUCCESS) {
+			UT_PRT("[AI_ISP] CVI_VI_AiIspInfo PUT failed, ret=%#x\n", s32Ret);
+		}
+
+		loop_cnt++;
+		if (loop_cnt >= AI_ISP_TEST_LOOP_CNT) {
+			UT_PRT("[AI_ISP] reached test loop count %u, stopping\n", loop_cnt);
+			break;
+		}
+	}
+
+	ai_isp_thread_running = CVI_FALSE;
+	UT_PRT("[AI_ISP] raw loop thread end, pipe=%d, loops=%u\n", ViPipe, loop_cnt);
+	return NULL;
+}
+
+CVI_S32 vi_ut_ai_isp_test(VI_UT_CTX *pUtCtx)
+{
+	CVI_S32 s32Ret = CVI_SUCCESS;
+	VI_PIPE ViPipe = 0;
+	VI_AI_ISP_CFG_S ai_isp_cfg;
+
+	UT_PRT("[AI_ISP] test start\n");
+
+	s32Ret = vi_ut_sys_init(pUtCtx);
+	if (s32Ret != CVI_SUCCESS) {
+		UT_PRT("[AI_ISP] sys init failed, ret=%#x\n", s32Ret);
+		return s32Ret;
+	}
+
+	s32Ret = vi_ut_set_vi_vpss_mode(pUtCtx);
+	if (s32Ret != CVI_SUCCESS) {
+		UT_PRT("[AI_ISP] set vi_vpss mode failed, ret=%#x\n", s32Ret);
+		return s32Ret;
+	}
+
+	// Step 1: AI ISP INIT
+	memset(&ai_isp_cfg, 0, sizeof(ai_isp_cfg));
+	ai_isp_cfg.viPipe = ViPipe;
+	ai_isp_cfg.enAiIspType = VI_AI_ISP_CFG_INIT;
+	ai_isp_cfg.reserved[0] = 1; // fp16
+
+	s32Ret = CVI_VI_AiIspCfg(&ai_isp_cfg);
+	if (s32Ret != CVI_SUCCESS) {
+		UT_PRT("[AI_ISP] INIT failed, ret=%#x\n", s32Ret);
+		goto ai_isp_test_deinit;
+	}
+	UT_PRT("[AI_ISP] INIT success\n");
+
+	s32Ret = vi_ut_vi_init(pUtCtx);
+	if (s32Ret != CVI_SUCCESS) {
+		UT_PRT("[AI_ISP] vi init failed, ret=%#x\n", s32Ret);
+		return s32Ret;
+	}
+
+	// Step 2: AI ISP ENABLE
+	memset(&ai_isp_cfg, 0, sizeof(ai_isp_cfg));
+	ai_isp_cfg.viPipe = ViPipe;
+	ai_isp_cfg.enAiIspType = VI_AI_ISP_CFG_ENABLE;
+
+	s32Ret = CVI_VI_AiIspCfg(&ai_isp_cfg);
+	if (s32Ret != CVI_SUCCESS) {
+		UT_PRT("[AI_ISP] ENABLE failed, ret=%#x\n", s32Ret);
+		goto ai_isp_test_deinit_vi;
+	}
+	UT_PRT("[AI_ISP] ENABLE success %d\n", ai_isp_cfg.enAiIspType);
+
+	// Step 3: start raw loop thread (get -> bypass memcpy -> put)
+	ai_isp_thread_running = CVI_TRUE;
+	s32Ret = pthread_create(&ai_isp_thread_id, NULL,
+		ai_isp_raw_loop_thread, (void *)(uintptr_t)ViPipe);
+	if (s32Ret != 0) {
+		UT_PRT("[AI_ISP] create thread failed, ret=%d\n", s32Ret);
+		ai_isp_thread_running = CVI_FALSE;
+		goto ai_isp_test_disable;
+	}
+
+	// wait for thread to finish
+	pthread_join(ai_isp_thread_id, NULL);
+	UT_PRT("[AI_ISP] raw loop thread joined\n");
+
+ai_isp_test_disable:
+	// Step 4: AI ISP DISABLE
+	memset(&ai_isp_cfg, 0, sizeof(ai_isp_cfg));
+	ai_isp_cfg.viPipe = ViPipe;
+	ai_isp_cfg.enAiIspType = VI_AI_ISP_CFG_DISABLE;
+
+	s32Ret = CVI_VI_AiIspCfg(&ai_isp_cfg);
+	if (s32Ret != CVI_SUCCESS) {
+		UT_PRT("[AI_ISP] DISABLE failed, ret=%#x\n", s32Ret);
+	} else {
+		UT_PRT("[AI_ISP] DISABLE success\n");
+	}
+
+ai_isp_test_deinit_vi:
+	// Step 5: AI ISP DEINIT
+	memset(&ai_isp_cfg, 0, sizeof(ai_isp_cfg));
+	ai_isp_cfg.viPipe = ViPipe;
+	ai_isp_cfg.enAiIspType = VI_AI_ISP_CFG_DEINIT;
+
+	s32Ret = CVI_VI_AiIspCfg(&ai_isp_cfg);
+	if (s32Ret != CVI_SUCCESS) {
+		UT_PRT("[AI_ISP] DEINIT failed, ret=%#x\n", s32Ret);
+	} else {
+		UT_PRT("[AI_ISP] DEINIT success\n");
+	}
+
+	return s32Ret;
+
+ai_isp_test_deinit:
+	s32Ret = vi_ut_vi_deinit(pUtCtx);
+	if (s32Ret != CVI_SUCCESS) {
+		UT_PRT("[AI_ISP] vi deinit failed, ret=%#x\n", s32Ret);
+	}
+
+	vi_ut_sys_exit();
+	UT_PRT("[AI_ISP] test end\n");
+
+	return s32Ret;
+}
diff --git a/self_test/vi_ut/vi_ut_comm.h b/self_test/vi_ut/vi_ut_comm.h
index a029cb8..6f490b8 100644
--- a/self_test/vi_ut/vi_ut_comm.h
+++ b/self_test/vi_ut/vi_ut_comm.h
@@ -185,5 +185,6 @@ CVI_S32 vi_ut_vpss_deinit(VI_UT_CTX *pUtCtx);
 CVI_S32 vi_ut_vi_deinit(VI_UT_CTX *pUtCtx);
 CVI_VOID vi_ut_sys_exit();
 CVI_S32 vi_ut_set_vi_vpss_mode(VI_UT_CTX *pUtCtx);
+CVI_S32 vi_ut_ai_isp_test(VI_UT_CTX *pUtCtx);
 
 #endif
diff --git a/self_test/vi_ut/vi_ut_main.c b/self_test/vi_ut/vi_ut_main.c
index 2c52d03..21b3011 100644
--- a/self_test/vi_ut/vi_ut_main.c
+++ b/self_test/vi_ut/vi_ut_main.c
@@ -47,6 +47,7 @@ enum CASE_UT {
 	CASE_VI_MULTI_INIT_TEST,
 	// CASE_VI_PLD_TEST,
 	// CASE_VI_SLT_TEST,
+	CASE_AI_ISP_TEST,
 	CASE_MAX,
 };
 
@@ -1107,6 +1108,25 @@ static int case_vi_multi_init_test(void *p)
 	return s32Ret;
 }
 
+static int case_ai_isp_test(void *p)
+{
+	CVI_S32 s32Ret = CVI_SUCCESS;
+	VI_UT_CTX *pUtCtx = (VI_UT_CTX *)p;
+
+	pUtCtx->viVpssMode = VI_OFFLINE_VPSS_OFFLINE;
+	pUtCtx->isDpcmOn = true;
+
+	scanf_fe_dram_post_dram_info(pUtCtx);
+
+	s32Ret = vi_ut_ai_isp_test(pUtCtx);
+	if (s32Ret != CVI_SUCCESS) {
+		UT_PRT("vi_ut_ai_isp_test failed. s32Ret: 0x%x !\n", s32Ret);
+		return s32Ret;
+	}
+
+	return s32Ret;
+}
+
 static const struct vi_ut_info vi_uts[] = {
 	UT_INFO(CASE_RAW_REPLAY,			case_raw_replay,		0),
 	UT_INFO(CASE_SENSOR_ON_THE_FLY,			case_sensor_onthefly,		0),
@@ -1130,6 +1150,7 @@ static const struct vi_ut_info vi_uts[] = {
 	UT_INFO(CASE_LOAD_MESH_LDC,			case_load_mesh_ldc,		0),
 	UT_INFO(CASE_VI_SDK_TEST,			case_vi_sdk_test,		0),
 	UT_INFO(CASE_VI_MULTI_INIT_TEST,		case_vi_multi_init_test,	0),
+	UT_INFO(CASE_AI_ISP_TEST,			case_ai_isp_test,		0),
 };
 
 static const CVI_U8 *strlwc(const CVI_U8 *in, CVI_U8 *out, CVI_U32 len)
-- 
2.25.1

